README.md
# Coloring verifier

Coloring-verifier is a small program to verify solutions to graph coloring problems in
the context of the DIMACS Computational Challenges.  

Two input files are
required: the instance file in the ASCII DIMACS file format (see
http://mat.gsia.cmu.edu/COLOR04/ for details), and the solution file.

The solution file has to be an ASCII file consisting of a single column
of numbers. Each number represents the color assigned to the node
numbered as the corresponding row number. Both colors and nodes are
assumed to start with 1. 

A directory called "examples" is included in the archive with an
example of instance and two solutions: one feasible and another
infeasible for the classical vertex graph coloring problem.

In case a solution is not feasible the program returns the number of
constraint violations. This number is computed as number of conflicting
edges, that is, edges with ending nodes having the same
color. Furthermore, it also returns the number of nodes involved in at
least one conflict.

The verifier can be used for the following problems:

1. vertex graph coloring 
2. list coloring
3. set coloring
4. T coloring
5. set-T-coloring. 
6. interval coloring
7. sum coloring


## BUILDING

The code has been tested on Linux and MacOsX operating systems. For
compiling the source code in both environments a call to the
Makefile by the `make` command should suffice.



See test_sol --help for more details.


---
Marco Chiarandini, 
University of Southern Denmark
email: marco@imada.sdu.dk

changes.md
## Version 3.0

- 06 Sept 2022. Moved to git.

- 20 May 2020. Contribution by Willem van Hoeve. Updated test_sol.c to scan for "c "
  and "p " (with a space) to avoid that when parsing the dimension of
  DIMACS instances, it scans for a "c" in the line, even if this "c" may
  appear anywhere.  Since some instances have "p col" for the dimension,
  the program was skipping that line as it contains a "c".  Moreover, it
  has been added the cases "edge", "edges", and "col" to identify the
  possible edge dimension. 



## Previous versions 

Available at: https://imada.sdu.dk/~marco/gcp/gcp_check_sol

| Version               | Date               | Note                                                                                                  |
| --------------------- | ------------------ | ----------------------------------------------------------------------------------------------------- |
| gcp_check_sol-2.8.tgz | 2018-10-29 19:50   | Bug fix: fixed the reporting of the number of edges from an ascii instance. (Thanks to Marko Bukal)   |
| gcp_check_sol-2.7.tgz | 2013-10-23 19:04   | Handling of empty or incomplete colorings in solution file                                            |
| gcp_check_sol-2.6.tgz | 2013-10-14 09:50   | Added reader for bin instance file format                                                             |
| test_sol-2.5.tgz      | 2013-08-06 19:03   | Added sum coloring calculation                                                                        |
| test_sol-2.4.tgz      | 2012-10-04 17:27   | Added interval coloring checker                                                                       |
| test_sol: V2.3        | 2010-10-06 16:22   | corrected a bug in memory allocation                                                                  |
| test_sol:  V2.2       | 2004-02-10 16:50   |                                                                                                       |



src/test_sol.c
/*****************************************************************************/
/*                                                                           */
/* Version:  3.0   Date:  06/09/2022   File: test_sol.c                      */
/* Author:  Marco Chiarandini                                                */
/* email: marco@imada.sdu.dk                                                 */
/*                                                                           */
/*****************************************************************************/

#include "test_sol.h"

#ifndef STR_ERR_UNKNOWN_LONG_OPT
# define STR_ERR_UNKNOWN_LONG_OPT   "%s: unrecognized option `--%s'\n"
#endif

#ifndef STR_ERR_LONG_OPT_AMBIGUOUS
# define STR_ERR_LONG_OPT_AMBIGUOUS "%s: option `--%s' is ambiguous\n"
#endif

#ifndef STR_ERR_MISSING_ARG_LONG
# define STR_ERR_MISSING_ARG_LONG   "%s: option `--%s' requires an argument\n"
#endif

#ifndef STR_ERR_UNEXPEC_ARG_LONG
# define STR_ERR_UNEXPEC_ARG_LONG   "%s: option `--%s' doesn't allow an argument\n"
#endif

#ifndef STR_ERR_UNKNOWN_SHORT_OPT
# define STR_ERR_UNKNOWN_SHORT_OPT  "%s: unrecognized option `-%c'\n"
#endif

#ifndef STR_ERR_MISSING_ARG_SHORT
# define STR_ERR_MISSING_ARG_SHORT  "%s: option `-%c' requires an argument\n"
#endif

#define STR_HELP_INSTANCE			\
  "  -i, --instance ARG    the instance file\n"

#define STR_HELP_SOLUTION			\
  "  -s, --solution ARG    the solution file\n"

#define STR_HELP_PROBLEM				\
  "  -p, --problem ARG     the problem -- see help\n"

#define STR_HELP_HELP						\
  "      --help            display this help text and exit\n"

#define STR_HELP				\
  STR_HELP_INSTANCE				\
  STR_HELP_SOLUTION				\
  STR_HELP_PROBLEM				\
  STR_HELP_HELP

/* Set to 1 if option --instance (-i) has been specified.  */
char opt_instance;

/* Set to 1 if option --solution (-s) has been specified.  */
char opt_solution;

/* Set to 1 if option --problem (-t) has been specified.  */
char opt_problem;

/* Set to 1 if option --help has been specified.  */
char opt_help;

/* Parse command line options.  Return index of first non-option argument,
   or -1 if an error is encountered.  */

/*************************************************************
 ************** utilities part ********************************
 ************************************************************/

void usage() {
  fprintf(stderr, "\nSolution Verifier for Graph Coloring, V2.6\n\n");
  fprintf(stderr,
	  "Usage: test_sol -i [INST_FILE] -s [SOL_FILE] -p [PROB_TYPE]\n");
  fprintf(stderr,
	  "--help            -h  \t help: prints this information\n\n");
  fprintf(stderr,
	  "Instance File Format: DIMACS file format. Nodes start from 1.\n");
  fprintf(stderr,
	  "Solution File Format: colors assigned to nodes wrote in column.\n");
  fprintf(stderr,
	  "After each entry the character \\n (new line) has to be printed.\n");
  fprintf(stderr,
	  "Colors start from 1 and the first color in the column represent\n");
  fprintf(stderr, "the color assigned to node 1. \n\n");
  fprintf(stderr,
	  "The Problem: coloring 1, list coloring 2, set coloring 3,\n");
  fprintf(stderr,
	  "             T-coloring 4, set T-coloring 5, interval-col 6\n");
  fprintf(stderr, "Report bugs to <marco@imada.sdu.dk>\n\n");
}

void checkOptions(const char * const program_name, const int argc,
		  char ** const argv) {
  int r = parse_options(program_name, argc, argv);

  if (r < 0) {
    fprintf(stderr,
	    "test_sol: You must specify input files and type of problem\n");
    fprintf(stderr, "Try `test_sol --help' for more information.\n");
    exit(1);
  }
  if (opt_help) {
    usage();
    exit(1);
  }
  if (!opt_instance || !opt_solution || !opt_problem) {
    fprintf(stderr,
	    "test_sol: You must specify input files and type of problem\n");
    fprintf(stderr, "Try `test_sol --help' for more information.\n");
    exit(1);
  }
}

int parse_options(const char * const program_name, const int argc,
		  char ** const argv) {
  static const char * const optstr__instance = "instance";
  static const char * const optstr__solution = "solution";
  static const char * const optstr__problem = "problem";
  static const char * const optstr__help = "help";
  int i = 0;
  opt_instance = 0;
  opt_solution = 0;
  opt_problem = 0;
  opt_help = 0;
  instance_name = 0;
  solution_name = 0;
  problem = 0;
  while (++i < argc) {
    const char *option = argv[i];
    if (*option != '-')
      return i;
    else if (*++option == '\0')
      return i;
    else if (*option == '-') {
      const char *argument;
      size_t option_len;
      ++option;
      if ((argument = strchr(option, '=')) == option)
	goto error_unknown_long_opt;
      else if (argument == 0)
	option_len = strlen(option);
      else
	option_len = argument++ - option;
      switch (*option) {
      case '\0':
	return i + 1;
      case 'h':
	if (strncmp(option + 1, optstr__help + 1, option_len - 1)
	    == 0) {
	  if (argument != 0) {
	    option = optstr__help;
	    goto error_unexpec_arg_long;
	  }
	  opt_help = 1;
	  return i + 1;
	}
	goto error_unknown_long_opt;
      case 'i':
	if (strncmp(option + 1, optstr__instance + 1, option_len - 1)
	    == 0) {
	  if (argument != 0)
	    instance_name = argument;
	  else if (++i < argc)
	    instance_name = argv[i];
	  else {
	    option = optstr__instance;
	    goto error_missing_arg_long;
	  }
	  opt_instance = 1;
	  break;
	}
	goto error_unknown_long_opt;
      case 'p':
	if (strncmp(option + 1, optstr__problem + 1, option_len - 1)
	    == 0) {
	  if (argument != 0)
	    problem = atoi(argument);
	  else if (++i < argc)
	    problem = atoi(argv[i]);
	  else {
	    option = optstr__problem;
	    goto error_missing_arg_long;
	  }
	  opt_problem = 1;
	  break;
	}
	goto error_unknown_long_opt;
      case 's':
	if (strncmp(option + 1, optstr__solution + 1, option_len - 1)
	    == 0) {
	  if (argument != 0)
	    solution_name = argument;
	  else if (++i < argc)
	    solution_name = argv[i];
	  else {
	    option = optstr__solution;
	    goto error_missing_arg_long;
	  }
	  opt_solution = 1;
	  break;
	}
      default:
      error_unknown_long_opt: fprintf(stderr,
				      STR_ERR_UNKNOWN_LONG_OPT, program_name, option);
	return -1;
      error_missing_arg_long: fprintf(stderr,
				      STR_ERR_MISSING_ARG_LONG, program_name, option);
	return -1;
      error_unexpec_arg_long: fprintf(stderr,
				      STR_ERR_UNEXPEC_ARG_LONG, program_name, option);
	return -1;
      }
    } else
      do {
	switch (*option) {
	case 'i':
	  if (option[1] != '\0')
	    instance_name = option + 1;
	  else if (++i < argc)
	    instance_name = argv[i];
	  else
	    goto error_missing_arg_short;
	  option = "\0";
	  opt_instance = 1;
	  break;
	case 's':
	  if (option[1] != '\0')
	    solution_name = option + 1;
	  else if (++i < argc)
	    solution_name = argv[i];
	  else
	    goto error_missing_arg_short;
	  option = "\0";
	  opt_solution = 1;
	  break;
	case 'p':
	  if (option[1] != '\0')
	    problem = atoi(option + 1);
	  else if (++i < argc)
	    problem = atoi(argv[i]);
	  else
	    goto error_missing_arg_short;
	  option = "\0";
	  opt_problem = 1;
	  break;
	default:
	  fprintf(stderr, STR_ERR_UNKNOWN_SHORT_OPT, program_name,
		  *option);
	  return -1;
	error_missing_arg_short: fprintf(stderr,
					 STR_ERR_MISSING_ARG_SHORT, program_name, *option);
	  return -1;
	}
      } while (*++option != '\0');
  }
  return i;
}

long int **allocateMatrix(long int rows, long int cols) {
  long int i;
  long int **matrix;

  matrix = (long int **) malloc((rows + 1) * sizeof(long int *));
  if (matrix == NULL ) {
    fprintf(stderr, "Problems in memory allocation\n");
  }
  for (i = 0; i < (rows + 1); i++) {
    matrix[i] = (long int *) malloc((cols + 1) * sizeof(long int));
    if (matrix[i] == NULL ) {
      fprintf(stderr, "Problems in memory allocation\n");
    }
    memset(matrix[i], 0, (cols + 1) * sizeof(long int));
  }
  return matrix;
}

long int *allocateLongVector(long int size) {
  long int *vector;

  vector = (long int *) malloc((size + 1) * sizeof(long int));
  if (vector == NULL ) {
    fprintf(stderr, "Problems in memory allocation\n");
  }
  memset(vector, 0, (size + 1) * sizeof(long int));
  return vector;
}

int *allocateIntVector(int size) {
  int *vector;

  vector = (int *) malloc((size + 1) * sizeof(int));
  if (vector == NULL ) {
    fprintf(stderr, "Problems in memory allocation\n");
  }
  memset(vector, 0, (size + 1) * sizeof(int));
  return vector;
}

int int_compare(const void * a, const void * b) {
  return (*(int*) a - *(int*) b);
}

/*************************************************************
 **************** checker part ********************************
 ************************************************************/
void checkConstraints_bin() {
  long int v = 0, w = 0, col = 0;
  int assigned;
  int conflict = 0;
  long int *conflicting_nodes = NULL;

  int i, j;
  int dist, size;

  read_graph_DIMACS_bin(instance_name);
  //write_graph_DIMACS_ascii("tmp.col");
  conflicting_nodes = allocateLongVector(NUM_NODES);
  solution_sizes = allocateIntVector(NUM_NODES);
  solution_sets = (int **) malloc((NUM_NODES + 1) * sizeof(int *));
  assert(solution_sets != NULL);
  for (i = 0; i <= NUM_NODES; i++) {
    solution_sets[i] = (int *) malloc(sizeof(int));
  }

  readSolution();
  CONF_EDGES = 0;
  CONF_NODES = 0;
  NOT_IN_LIST = 0;


  for (v = 1; v <= NUM_NODES - 1; v++) {
    for (w = v + 1; w <= NUM_NODES; w++)
      if (get_edge(w-1, v-1)) {
	//printf("e %d %d\n",w+1,v+1 );
	conflict = 0;
	for (i = 0; i < solution_sizes[v]; i++) {
	  for (j = 0; j < solution_sizes[w]; j++) {
	    if (solution_sets[v][i] == solution_sets[w][j]) {
	      conflict++;
	      if (VERBOSE)
		printf("%ld %ld\n", v, w);
	    }
	  }
	}
	if (conflict > 0) {
	  for (i = 0; i < CONF_NODES; i++) {
	    if (conflicting_nodes[i] == v)
	      break;
	  }
	  if (i == CONF_NODES) {
	    conflicting_nodes[i] = v;
	    CONF_NODES++;
	  }
	  for (i = 0; i < CONF_NODES; i++) {
	    if (conflicting_nodes[i] == w)
	      break;
	  }
	  if (i == CONF_NODES) {
	    conflicting_nodes[i] = w;
	    CONF_NODES++;
	  }
	  CONF_EDGES++;
	}
      }
  }
  free(conflicting_nodes);
}

void checkConstraints_ascii() {
  FILE *col_file;
  char buf[LINE_BUF_LEN];
  char *p, *tmp;
  long int from = 0, to = 0, col = 0;
  int assigned;
  int conflict = 0;
  long int *conflicting_nodes = NULL;

  int i, j;
  int dist, size;


  col_file = fopen(instance_name, "r");
  assert(col_file != NULL);

  while (!feof(col_file)) {

    buf[0] = '\0';
    fgets((char*) &buf, LINE_BUF_LEN, col_file);
    if ((p = strstr((char*) (&buf), "c ")) != NULL ) {
      /* If a c is found in a line, it's part of header */
      printf("%s", buf);
    } else if ((p = strstr((char*) (&buf), "p ")) != NULL ) {
      /* This is the line with the problem dimension */
      printf("%s", buf);
      /*	    printf("reading parameter line ...\n");*/
      if (strstr((char*) &buf, "edge") != NULL
	  || (strstr((char*) &buf, "col") != NULL)
	  || (strstr((char*) &buf, "edges") != NULL)
	  || strstr((char*) &buf, "band") != NULL) {
	p++;
	while (*p != '\0' && *p != '\n') {
	  while (*p == ' ')
	    p++;
	  while (*p != ' ')
	    p++;
	  while (*p == ' ')
	    p++;
	  NUM_NODES = atol(p);
	  
	  while (*p != ' ') p++;
	  while (*p == ' ')
	    p++;
	  NUM_EDGES = atol(p);
	  conflicting_nodes = allocateLongVector(NUM_NODES);

	  solution_sizes = allocateIntVector(NUM_NODES);
	  solution_sets = (int **) malloc(
					  (NUM_NODES + 1) * sizeof(int *));
	  assert(solution_sets != NULL);
	  for (i = 0; i <= NUM_NODES; i++) {
	    solution_sets[i] = (int *) malloc(sizeof(int));
	  }

	  readSolution();
	  CONF_EDGES = 0;
	  CONF_NODES = 0;
	  NOT_IN_LIST = 0;
	  while (*p != ' ')
	    p++;
	  while (*p == ' ')
	    p++;
	  *p = '\0';
	}
      }
    } else if (buf[0] == 'e') /*((p=strstr((char*)(&buf), "e"))!=NULL) */
      { /* This is a line with an edge */

	tmp = strtok(buf, " ");

	tmp = strtok(0, " ");

	from = atoi(tmp);
	tmp = strtok(0, " ");

	to = atoi(tmp);

	tmp = strtok(0, " ");

	if (tmp != NULL ) {
	  dist = atoi(tmp);
	} else
	  dist = 1;
	
	assert((1 <= from) && (from <= NUM_NODES));
	assert((1 <= to) && (to <= NUM_NODES));
	conflict = 0;
	if (from == to) {
	  if (problem == 5)
	    for (i = 0; i < solution_sizes[from] - 1; i++)
	      for (j = i + 1; j < solution_sizes[to]; j++)
		if (abs(
			solution_sets[from][i]
			- solution_sets[to][j]) < dist) {
		  conflict++;
		  if (VERBOSE)
		    printf("%ld %d %d %d %d %d\n", from, i, j,
			   solution_sets[from][i],
			   solution_sets[from][j], dist);
		}
	} else {

	  for (i = 0; i < solution_sizes[from]; i++)
	    for (j = 0; j < solution_sizes[to]; j++)
	      if (problem == 4 || problem == 5) {
		if (abs(
			solution_sets[from][i]
			- solution_sets[to][j]) < dist) {
		  conflict++;
		  if (VERBOSE)
		    printf("%d %d (%ld,%d) (%ld,%d) %d\n", i, j,
			   from, solution_sets[from][i], to,
			   solution_sets[to][j], dist);
		}
	      } else {
		if (solution_sets[from][i]
		    == solution_sets[to][j]) {
		  conflict++;
		  if (VERBOSE)
		    printf("%ld %ld\n", from, to);
		}
	      }
	}
	if (conflict > 0) {
	  for (i = 0; i < CONF_NODES; i++) {
	    if (conflicting_nodes[i] == from)
	      break;
	  }
	  if (i == CONF_NODES) {
	    conflicting_nodes[i] = from;
	    CONF_NODES++;
	  }
	  for (i = 0; i < CONF_NODES; i++) {
	    if (conflicting_nodes[i] == to)
	      break;
	  }
	  if (i == CONF_NODES) {
	    conflicting_nodes[i] = to;
	    CONF_NODES++;
	  }
	  CONF_EDGES++;
	}
      } else if (buf[0] == 'n'
		 && (problem == 3 || problem == 5 || problem == 6)) {
      tmp = strtok(buf, " ");
      tmp = strtok(0, " ");
      from = atoi(tmp);
      tmp = strtok(0, " ");
      size = atoi(tmp);
      if (solution_sizes[from] != size) {
	SET_SIZE++;
      }
      if (problem == 6) {
	qsort(solution_sets[from], solution_sizes[from], sizeof(int),
	      int_compare);
	for (i = 0; i < solution_sizes[from]; i++)
	  printf("%d ", solution_sets[from][i]);
	printf("\n");
	for (i = 0; i < solution_sizes[from] - 1; i++)
	  if (solution_sets[from][i + 1] - solution_sets[from][i]
	      != 1) {
	    conflict++;
	    if (VERBOSE)
	      printf(
		     "non consecutive integers to vertex %ld: (%d %d) of ",
		     from, solution_sets[from][i + 1],
		     solution_sets[from][i]);
	    for (i = 0; i < solution_sizes[from]; i++)
	      printf("%d ", solution_sets[from][i]);
	    printf("\n");
	  }

      }
    } else if (buf[0] == 'f' && problem == 2) {
      tmp = strtok(buf, " ");
      tmp = strtok(0, " ");

      from = atoi(tmp);
      tmp = strtok(0, " ");
      assigned = 0;

      while (tmp != NULL ) {
	col = atoi(tmp);

	for (i = 0; i < solution_sizes[from]; i++)
	  if (solution_sets[from][i] == col) {
	    assigned = 1;
	  }
	tmp = strtok(0, " ");
      }
      if (assigned != 1) {
	NOT_IN_LIST++;
	if (VERBOSE)
	  printf("<---%ld %ld\n", from, col);
      }
    }

  }
  fclose(col_file);
  free(conflicting_nodes);
}

char checkConstraints() {
  int i;

  NUM_EDGES = 0;
  CONF_EDGES = 0;
  CONF_NODES = 0;
  NOT_IN_LIST = 0;
  SET_SIZE = 0;

  printf("\nreading coloring-problem-file %s ... \n\n", instance_name);

  int format = 0;

  if (strstr(instance_name, ".b\0") != NULL )
    format = BIN_FORMAT;
  else
    format = ASCII_FORMAT;

  if (format == BIN_FORMAT) {
    printf("...read input in bin format...\n");
    checkConstraints_bin();
  } else if (format == ASCII_FORMAT) {
    printf("...read input in ASCII format...");
    checkConstraints_ascii();
  }

  free(solution_sizes);
  for (i = 0; i < NUM_NODES + 1; i++)
    free(solution_sets[i]);
  free(solution_sets);

  DENSITY = ((float) (NUM_EDGES)) / ((NUM_NODES / 2) * (NUM_NODES - 1));

  if (COLORED_NODES==NUM_NODES && CONF_EDGES == 0 && NOT_IN_LIST == 0 && SET_SIZE == 0)
    return TRUE;
  else
    return FALSE;
}

void readSolution() {
  FILE *sol_file;
  long int node, count;
  long int col;
  char *p;
  long int *colors = allocateLongVector(NUM_NODES * 20);
  long int i;
  char buf[LINE_BUF_LEN];
  sol_file = fopen(solution_name, "r");
  assert(sol_file != NULL);

  node = 0;
  COLORED_NODES=0;
  NUM_COL = 0;
  SUM_COL = 0;
  MIN_COL = INT_MAX;
  MAX_COL = 0;
  char new_line;
  
  
  while (fgets((char*) &buf, LINE_BUF_LEN, sol_file)!=NULL) {
	 //!feof(sol_file)) {
    count = 0;
    new_line=TRUE;
    p = strtok(buf, " ");

    do {

      if (*p != '\n') {
	if (new_line) {
	  COLORED_NODES++;
	  node++;
	  new_line=FALSE;
	}
	col = atoi(p);
	solution_sets[node] = realloc(solution_sets[node],
				      (count + 1) * sizeof(int));
	solution_sets[node][count++] = col;
	for (i = 0; i < NUM_COL; i++) {
	  if (colors[i] == col)
	    break;
	}
	if (i == NUM_COL) {
	  colors[i] = col;
	  NUM_COL++;
	}
	if (col < MIN_COL)
	  MIN_COL = col;
	if (col > MAX_COL)
	  MAX_COL = col;
	SUM_COL += col;
      }
      p = strtok(0, " ");
    } while (p != NULL );
    solution_sizes[node] = count;
    if (count < 1)
      SET_SIZE++;

    
    //printf("%d %d %d %d\n",NUM_NODES,node,COLORED_NODES, col);
    //    fgets((char*) &buf, LINE_BUF_LEN, sol_file);
    //printf("%s",buf);
  }

  fclose(sol_file);
}



src/test_sol.h
/*****************************************************************************/
/*                                                                           */
/* Version:  3.0    Date:  19/02/2004-06/09/2022   File: test_sol.h          */
/* Author:  Marco Chiarandini                                                */
/* email: machud@intellektik.informatik.tu-darmstadt.de                      */
/*                                                                           */
/*****************************************************************************/

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>


#define LINE_BUF_LEN     2000

#define VERBOSE 1

#ifdef NOGETOPT
static char name_buf[LINE_BUF_LEN];
static char sol_buf[LINE_BUF_LEN];
//static int     opt;
#endif

#define TRUE 1
#define FALSE 0

#define ASCII_FORMAT	1
#define BIN_FORMAT	2

#define MAX_PREAMBLE 10000

/* If you change MAX_NR_VERTICES, change MAX_NR_VERTICESdiv8 to be
the 1/8th of it */
//#define NMAX 10000	/* maximum number of vertices handles */
#define MAX_NR_VERTICES		5000	/* = NMAX */
#define MAX_NR_VERTICESdiv8	625 	/* = NMAX/8 */

extern int **solution_sets;
extern int *solution_sizes;

extern  long int NUM_COL;
extern  long int MAX_COL;
extern  long int MIN_COL;
extern  long int NUM_EDGES;
extern  long int NUM_NODES;
extern float DENSITY;
extern long int COLORED_NODES;
extern  long int CONF_EDGES;
extern  long int CONF_NODES;
extern  long int NOT_IN_LIST;
extern  long int SET_SIZE;
extern  long int SUM_COL;

extern const char *instance_name;
extern const char *solution_name;
extern int problem;


char Bitmap[MAX_NR_VERTICES][MAX_NR_VERTICESdiv8];
char Preamble[MAX_PREAMBLE];

void usage();
void checkOptions(const char *const program_name, const int argc, char **const argv);
int parse_options (const char *const program_name, const int argc, char **const argv);

int get_params(void);
char get_edge(  int i,  int j );
void read_graph_DIMACS_bin(const char *file);
void write_graph_DIMACS_ascii(const char *file);


char checkConstraints();
void checkConstraints_ascii();
void checkConstraints_bin();
void readSolution();



src/bin2asc.c
/* Dimacs graph format translator to and from a binary, more efficient
   format. Written by Tamas Badics (badics@rutcor.rutgers.edu), 
   using the technique of Marcus Peinado, Boston University. */
/*
   Corrected to number nodes from 1 to n (not 0 to n-1)
   MT
*/

#include "test_sol.h"

const char masks[ 8 ] = { 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80 };


int get_params(void);
/* ====================================================== */

char get_edge(  int i,  int j )
{
	int byte, bit;
	char mask;
	
	bit  = 7-(j & 0x00000007);
	byte = j >> 3;
	
	mask = masks[bit];
	return( (Bitmap[i][byte] & mask)==mask );
}

/* ============================================= */
void write_graph_DIMACS_ascii(const char *file)
{
	int i,j;
	FILE *fp;

	if ( (fp=fopen(file,"w"))==NULL )
	  { printf("ERROR: Cannot open outfile\n"); exit(10); }

	fprintf(fp, "%s",Preamble);

	for ( i = 0; i<NUM_NODES; i++ )
	  {
		  for ( j=0; j<=i; j++ )
			if ( get_edge(i,j) ) fprintf(fp,"e %d %d\n",i+1,j+1 );
	  }

	fclose(fp);
}

void read_graph_DIMACS_bin(const char *file)
{

	int i, length = 0;
	FILE *fp;
	
	if ( (fp=fopen(file,"r"))==NULL )
	  { printf("ERROR: Cannot open infile\n"); exit(10); }

	if (!fscanf(fp, "%d\n", &length))
	  { printf("ERROR: Corrupted preamble.\n"); exit(10); }

	if(length >= MAX_PREAMBLE)
	  { printf("ERROR: Too long preamble.\n"); exit(10); }
		   
	fread(Preamble, 1, length, fp);
	Preamble[length] = '\0';
	
	if (!get_params())
		  { printf("ERROR: Corrupted preamble.\n"); exit(10); }

	for ( i = 0
		 ; i < NUM_NODES && fread(Bitmap[i], 1, (int)((i + 8)/8), fp)
		 ; i++ );

	fclose(fp);
}

int get_params(void)
                      /* getting Nr_vert and Nr_edge from the preamble string, 
						 containing Dimacs format "p ??? num num" */
{
	char c, *tmp;
	char * pp = Preamble;
	int stop = 0;
	tmp = (char *)calloc(100, sizeof(char));
	
	NUM_NODES = NUM_EDGES = 0;
	
	while (!stop && (c = *pp++) != '\0'){
		switch (c)
		  {
			case 'c':
			  while ((c = *pp++) != '\n' && c != '\0');
			  break;
			  
			case 'p':
			  sscanf(pp, "%s %ld %ld\n", tmp, &NUM_NODES, &NUM_EDGES);
			  stop = 1;
			  break;
			  
			default:
			  break;
		  }
	}
	
	free(tmp);
	
	if (NUM_NODES == 0 || NUM_EDGES == 0)
	  return 0;  /* error */
	else
	  return 1;
	
}


src/main.c
/*****************************************************************************/
/*                                                                           */
/* Version:  3.0   Date:  06/09/2022   File: main.c                          */
/* Author:  Marco Chiarandini                                                */
/* email: marco@imada.sdu.dk                                                 */
/*                                                                           */
/*****************************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include "test_sol.h"


int **solution_sets;
int *solution_sizes;
long int **constraints_matrix;
 long int NUM_COL;
 long int SUM_COL;
 long int MIN_COL;
 long int MAX_COL;
 long int NUM_EDGES;
 long int NUM_NODES;
float DENSITY;
 long int COLORED_NODES;
 long int CONF_NODES;
 long int CONF_EDGES;
 long int NOT_IN_LIST;
 long int SET_SIZE;
const char *instance_name;
const char *solution_name;
int problem;


int main( argc, argv )
     int     argc;
     char    *argv[];
{

  char proper;

  checkOptions("test_sol", argc, argv );

  proper = checkConstraints();
  /*num_of_viol = count_violations();*/
  
  
  printf("\ninstance: %s\n",instance_name);
  printf("solution: %s\n",solution_name);
  printf("nodes: %ld\n",NUM_NODES);
  printf("edges: %ld\n",NUM_EDGES);
  printf("density: %.3f\n",DENSITY);
  
  if (proper)
    {
      printf("+---------------------------------------------------------+\n");
      printf("| The proposed coloring uses %4ld colors                  |\n",NUM_COL);
      printf("| from %2ld to %4ld with a max spread (T-span) of %4ld      |\n",MIN_COL,MAX_COL, MAX_COL - MIN_COL);
      printf("| It has a sum of %4ld.                                   |\n",SUM_COL);
      printf("| It satisfies all the constraints of the instance!       |\n");
      printf("+---------------------------------------------------------+\n");   
    }
  else 
    {
      printf("+---------------------------------------------------------------------+\n");
      printf("| The proposed solution uses %4ld colors                              |\n",NUM_COL);
      printf("| from %2ld to %4ld with a max spread (T-span) of %4ld,                 |\n",MIN_COL,MAX_COL, MAX_COL - MIN_COL);
      printf("| It has a total sum of %4ld.                                         |\n",SUM_COL);
      printf("| However, it DOES NOT satisfy all the constraints of the instance!   |\n");
      printf("| There are:                                                          |\n");
      printf("| %4ld vertices not colored,                                          |\n",NUM_NODES-COLORED_NODES);
      printf("| %4ld constraints (edges) not satisfied,                             |\n",CONF_EDGES);
      printf("| %4ld nodes in conflict, and                                         |\n",CONF_NODES);
      printf("| %4ld nodes which do not receive the right n. of col.                |\n",SET_SIZE);
      printf("| The nodes that receive a colour not                                 |\n");
      printf("| in the preference list are %4ld                                     |\n",NOT_IN_LIST);
      printf("+---------------------------------------------------------------------+\n");
    }

  return EXIT_SUCCESS;
}
